jobserver/api/releases.py

```diff
@@ -79,3 +84,52 @@ def put(self, request, workspace_name, release_hash):
         )
         response["Release-Id"] = release.id
         return response
+
+
+def validate_release_access(request, release_hash):
+    try:
+        release = Release.objects.get(id=release_hash)
+    except Release.DoesNotExist:
+        raise NotFound(f"Release {release_hash} does not exist")
+
+    # TODO: check if release is published
+    if request.user.is_anonymous:
+        raise NotAuthenticated("Invalid user or token")
+    else:
+        # valid use
+        project = release.workspace.project
+        if not project.members.filter(username=request.user.username).exists():
+            raise NotAuthenticated(f"Invalid user or token for release {release_hash}")
+
+    return release
+
+
+class ReleaseIndexAPI(APIView):
+    def get(self, request, release_hash):
+        release = validate_release_access(request, release_hash)
+        return Response(release.files)
+
+
+class ReleaseFileAPI(APIView):
+    def get(self, request, release_hash, filepath):
+        release = validate_release_access(request, release_hash)
+        handle = release.file_path(filepath)
+
+        if handle is None:
+            raise NotFound(f"File {filepath} not found in release {release_hash}")
+
+        internal_redirect = request.headers.get("Releases-Redirect")
+        if internal_redirect:
+            # we're behind nginx, so use X-Accel-Redirect to serve the file from nginx
+            relative_path = handle.relative_to(settings.RELEASE_STORAGE)
+            redirect_path = f"{internal_redirect}/{relative_path}"
+            response = HttpResponse()
+            response.headers["X-Accel-Redirect"] = redirect_path
+        else:
+            # serve directly from django in dev
+            mtype, _ = mimetypes.guess_type(handle)
+            # use regular django response to bypass DRFs renderer framework and just serve bytes
+            response = HttpResponse(FileWrapper(handle.open()), content_type=mtype)
```

Comment:
Did you look at the FileResponse class for this?

Reply:
Great, used that, thanks.